src/CHIPBindings.cc

```diff
@@ -3700,25 +3719,26 @@ hipError_t hipMemcpy3DAsyncInternal(const struct hipMemcpy3DParms *Params,
       DstPtr = Params->dstPtr.ptr;
     }
   }
-  LOCK(ChipQueue->QueueMtx); // prevent interruptions
-  if ((WidthInBytes == DstPitch) && (WidthInBytes == SrcPitch)) {
-    return hipMemcpyInternal((void *)DstPtr, (void *)SrcPtr,
-                             WidthInBytes * Height * Depth, Params->kind);
-  } else {
+
+  if (YSize * XSize == 0)
+    return hipSuccess;
+
+  size_t srcSlicePitch = SrcPitch * YSize;
+  size_t dstSlicePitch = DstPitch * YSize;
+  if ((WidthInBytes == DstPitch) && (WidthInBytes == SrcPitch))
+    ChipQueue->memCopyAsync(DstPtr, SrcPtr, WidthInBytes * Height * Depth);
+  else {
+    // TODO What am I doing wrong here?
+    // ChipQueue->memCopy3DAsync(DstPtr, DstPitch, dstSlicePitch, SrcPtr, SrcPitch, srcSlicePitch, Width, Height, Depth);
```

Comment:
@pjaaskel  please take a look

Reply:
@franz you have better grasp of this API, perhaps an educated guess from you here?